File: problems/111-Minimum-Depth-of-Binary-Tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func traverse(root *TreeNode, current int) int {
	if root == nil {
		return current - 1
	} else if root.Left == nil && root.Right == nil {
		return current
	} else if root.Left != nil && root.Right == nil {
		return traverse(root.Left, current+1)
	} else if root.Left == nil && root.Right != nil {
		return traverse(root.Right, current+1)
	} else {
		left := traverse(root.Left, current+1)
		right := traverse(root.Right, current+1)

		if left <= right {
			return left
		} else {
			return right
		}
	}
}

func minDepth(root *TreeNode) int {
	return traverse(root, 1)
}

File: problems/111-Minimum-Depth-of-Binary-Tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int minDepth(TreeNode root) {
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        
        if (root != null) {
            queue.offer(root);
        }
        
        int result = 0;
        while (!queue.isEmpty()) {
            result++;

            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll();

                if (node.left == null && node.right == null) {
                    return result;
                } else if (node.left == null && node.right != null) {
                    queue.offer(node.right);
                } else if (node.left != null && node.right == null) {
                    queue.offer(node.left);
                } else {
                    queue.offer(node.left);
                    queue.offer(node.right);
                }
            }
        }

        return result;
    }
}

